File: sdk/tests/go-api/api/save-options.go
package api

import (
	"context"
	"encoding/json"
	"log"
	"net/http"
	"time"

	"github.com/selfxyz/self/sdk/tests/go-api/config"
)

type SaveOptionsRequest struct {
	UserID  string      `json:"userId"`
	Options interface{} `json:"options"`
}

type SaveOptionsResponse struct {
	Message string `json:"message"`
}

func SaveOptionsHandler(w http.ResponseWriter, r *http.Request) {
	// Enable CORS
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
	w.Header().Set("Access-Control-Allow-Headers", "Content-Type")

	if r.Method == "OPTIONS" {
		w.WriteHeader(http.StatusOK)
		return
	}

	if r.Method != "POST" {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusMethodNotAllowed)
		json.NewEncoder(w).Encode(map[string]string{"message": "Method not allowed"})
		return
	}

	w.Header().Set("Content-Type", "application/json")

	var req SaveOptionsRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]string{"message": "Invalid JSON"})
		return
	}

	if req.UserID == "" {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]string{"message": "User ID is required"})
		return
	}

	if req.Options == nil {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]string{"message": "Options are required"})
		return
	}

	// Initialize in-memory config store (replacing Redis)
	configStore, err := config.NewKVConfigStoreFromEnv()
	if err != nil {
		log.Printf("Failed to initialize config store: %v", err)
		w.WriteHeader(http.StatusInternalServerError)
		json.NewEncoder(w).Encode(map[string]string{"message": "Internal server error", "error": err.Error()})
		return
	}
	defer configStore.Close()

	// Store options in memory with 30-minute expiration (matching TypeScript: ex: 1800)
	ctx := context.Background()
	optionsJSON, err := json.Marshal(req.Options)
	if err != nil {
		log.Printf("Failed to marshal options: %v", err)
		w.WriteHeader(http.StatusInternalServerError)
		json.NewEncoder(w).Encode(map[string]string{"message": "Internal server error", "error": "Failed to serialize options"})
		return
	}

	// Use in-memory storage with expiration (30 minutes, matching TypeScript)
	err = configStore.SetWithExpiration(ctx, req.UserID, string(optionsJSON), 30*time.Minute)
	if err != nil {
		log.Printf("Failed to save options to storage: %v", err)
		w.WriteHeader(http.StatusInternalServerError)
		json.NewEncoder(w).Encode(map[string]string{"message": "Internal server error", "error": "Failed to save options"})
		return
	}

	log.Printf("Saved options for user: %s, options: %+v\n", req.UserID, req.Options)

	response := SaveOptionsResponse{
		Message: "Options saved successfully",
	}

	json.NewEncoder(w).Encode(response)
}

File: sdk/tests/ts-api/api/save-options.ts
import { Request, Response } from "express";

// In-memory storage for testing purposes (replaces Vercel KV)
const optionsStore = new Map<string, { data: string; expiry: number }>();

// Helper function to clean expired entries
const cleanExpiredEntries = () => {
  const now = Date.now();
  for (const [key, value] of optionsStore.entries()) {
    if (value.expiry < now) {
      optionsStore.delete(key);
    }
  }
};

export const saveOptionsHandler = async (
  req: Request,
  res: Response
) => {
  if (req.method !== "POST") {
    return res.status(405).json({ message: "Method not allowed" });
  }

  try {
    const { userId, options } = req.body;

    if (!userId) {
      return res.status(400).json({ message: "User ID is required" });
    }

    if (!options) {
      return res.status(400).json({ message: "Options are required" });
    }

    console.log("Saving options for user:", userId, options);

    // Store the options in memory with a 30-minute expiration
    const expiryTime = Date.now() + (30 * 60 * 1000); // 30 minutes from now
    optionsStore.set(userId, {
      data: JSON.stringify(options),
      expiry: expiryTime
    });

    // Clean expired entries periodically
    cleanExpiredEntries();

    return res.status(200).json({ message: "Options saved successfully" });
  } catch (error) {
    console.error("Error saving options:", error);
    return res.status(500).json({
      message: "Internal server error",
      error: error instanceof Error ? error.message : "Unknown error",
    });
  }
};

// Helper function to get options (used by verify handler)
export const getOptions = (userId: string): any => {
  cleanExpiredEntries();
  const stored = optionsStore.get(userId);
  if (stored && stored.expiry > Date.now()) {
    return JSON.parse(stored.data);
  }
  return null;
};
